runtime/caml/signals.h

```diff
@@ -34,14 +34,85 @@ CAMLextern void caml_enter_blocking_section (void); // raises
 CAMLextern void caml_enter_blocking_section_no_pending (void);
 CAMLextern void caml_leave_blocking_section (void);
 
+/* Asynchronous actions
+
+   We distinguish two kinds of asynchronous actions:
+
+   - non-raising actions, such as non-raising signal handlers (e.g.
+     systhread's yield).
+
+   - unrestricted asynchronous callbacks: possibly-raising signal
+     handlers, memprof callbacks and finalisers (including GC alarms).
+
+   Signal handlers in particular are defined as one of two kinds,
+   non-raising and raising. Non-raising handlers must guarantee that
+   they do not raise or fail, but also that they are brief (typically,
+   systhread's yield()), because themselves cannot be interrupted.
+
+   In contrast, asynchronous callbacks of the second kind are allowed
+   to raise an exception (typically, when handling SIGINT or SIGALRM),
+   are not limited in what they can express (typically, finalisers and
+   memprof callbacks), and can themselves be interrupted by
+   asynchronous callbacks.
+
+   Masking temporarily delays the execution of asynchronous callbacks.
+   There are two kinds of masking:
+
+   - The _nonpreemptible mask_ delays execution of all asynchronous
+     callbacks.
+
+   - The _uninterruptible mask_ delays the execution of unrestricted
+     asynchronous callbacks, whereas
+
+   This gives three possible values for
+   [Caml_state->mask_async_callbacks]: CAML_MASK_NONE,
+   CAML_MASK_UNINTERRUPTIBLE and CAML_MASK_NONPREEMPTIBLE.
```

Comment:
Beginner comment: to me "to interrupt" and "to preempt" are mostly synonymous, so I must say that I find the naming a bit confusing. Are you sure that it is very clear to more knowledgeable people? Would you maybe have alternate naming suggestions?

(For example I wondered if we could call short non-raising callbacks "short callbacks" or "small callbacks", and asynchronous callbacks "long callbacks" or "large callbacks", and whether that would suggest names here.)

Reply:
I now made sure to define _interrupt_ at the beginning of that documentation comment. In SML parlance, an interrupt is a special kind of exception that cancels of the current thread asynchronously.

I have plans to eventually name the last level differently if the notion of nonpreemptible mask does not make it.

I do not like small/long, which do not mean much. What matters is whether the current thread can be cancelled at that point (affine vs. linear), which is what raising vs. non-raising reflects to me.